rb_ws/src/buggy/buggy/buggy_state_converter.py

```diff
@@ -0,0 +1,165 @@
+        converted_msg.pose.pose.orientation.w = 0.0
+
+        # ---- 3. Copy Covariances (Unchanged) ----
+        # TODO: check whether covariances should be copied over or not sent at all
+        # converted_msg.pose.covariance = msg.pose.covariance
+        # converted_msg.twist.covariance = msg.twist.covariance
+
+        # ---- 4. Copy Linear Velocities ----
+        converted_msg.twist.twist.linear.x = msg.twist.twist.linear.x   # m/s in x-direction
+        converted_msg.twist.twist.linear.y = msg.twist.twist.linear.y   # m/s in x-direction
+        converted_msg.twist.twist.linear.z = msg.twist.twist.linear.z   # Keep original Z velocity (??)
+
+        # ---- 5. Copy Angular Velocities ----
+        converted_msg.twist.twist.angular.x = msg.twist.twist.angular.z   # rad/s for heading change rate (using yaw from twist.angular)
+        converted_msg.twist.twist.angular.y = 0.0   # ignored
+        converted_msg.twist.twist.angular.z = 0.0   # ignored
+
+        return converted_msg
+
+    def convert_NAND_state(self, msg):
+        """ Converts /NAND namespace raw state to clean state units and structure """
+        converted_msg = Odometry()
+        converted_msg.header = msg.header
+
+        # ---- 1. Directly use UTM Coordinates ----
+        converted_msg.pose.pose.position.x = msg.pose.pose.position.x   # UTM Easting
+        converted_msg.pose.pose.position.y = msg.pose.pose.position.y   # UTM Northing
+        converted_msg.pose.pose.position.z = 0.0                        # ignored
+
+        # ---- 2. Orientation in Radians with 0 at East ----
+        converted_msg.pose.pose.orientation.x = msg.pose.pose.orientation.x
+        # ignored:
+        converted_msg.pose.pose.orientation.y = 0.0
+        converted_msg.pose.pose.orientation.z = 0.0
+        converted_msg.pose.pose.orientation.w = 0.0
+
+        # ---- 3. Copy Covariances (Unchanged) ----
+        # TODO: check whether covariances should be copied over or not sent at all
+        # converted_msg.pose.covariance = msg.pose.covariance
+        # converted_msg.twist.covariance = msg.twist.covariance
+
+        # ---- 4. Copy Linear Velocities ----
+        # TODO: check that ROS serial translator node changes scalar speed to velocity x/y components before pushing to raw_state
+        converted_msg.twist.twist.linear.x = msg.twist.twist.linear.x   # m/s in x-direction
+        converted_msg.twist.twist.linear.y = msg.twist.twist.linear.y   # m/s in y-direction
+        converted_msg.twist.twist.linear.z = msg.twist.twist.linear.z   # Keep original Z velocity (??)
+
+        # ---- 5. Copy Angular Velocities ----
+        converted_msg.twist.twist.angular.x = msg.twist.twist.angular.x
+        converted_msg.twist.twist.angular.y = 0.0
+        converted_msg.twist.twist.angular.z = 0.0
+
+        return converted_msg
+
+    # replaced custom function with tf_transformations library function
+    def quaternion_to_yaw(self, qx, qy, qz, qw):
```

Comment:
appreciate you using your own function, but I'm unsure if it's correct. Can you instead use the function in this library `from tf.transformations import euler_from_quaternion`